packages/langium/src/workspace/document-builder.ts

```diff
@@ -114,58 +213,73 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
         });
     }
 
-    protected collectDocuments(changed: LangiumDocument[], deleted: URI[]): LangiumDocument[] {
-        const allUris = changed.map(e => e.uri).concat(deleted);
-        const affected = this.indexManager.getAffectedDocuments(allUris).toArray();
-        affected.forEach(e => {
-            const linker = this.serviceRegistry.getServices(e.uri).references.Linker;
-            linker.unlink(e);
-            e.state = Math.min(e.state, DocumentState.ComputedScopes); // need to re-index potentially linked references
-        });
-        const docSet = new Set([
-            ...changed,
-            ...affected,
-            // Also include all documents haven't completed the document lifecycle yet
-            ...this.langiumDocuments.all.filter(e => e.state < DocumentState.Validated)
-        ]);
-        return Array.from(docSet);
-    }
-
+    /**
+     * Build the given documents by stepping through all build phases. If a document's state indicates
+     * that a certain build phase is already done, the phase is skipped for that document.
+     */
     protected async buildDocuments(documents: LangiumDocument[], options: BuildOptions, cancelToken: CancellationToken): Promise<void> {
+        this.prepareBuild(documents, options);
         // 0. Parse content
```

Comment:
IMO the current numbering makes sense because it expresses that document of that state will be included in each phase. Example: documents with state Changed = 0 are included in the "Parse content" phase.